keps/sig-node/2837-pod-level-resource-spec/README.md

```diff
@@ -0,0 +1,1108 @@
+2. Support Pod-level resource requests/limits as an opt-in feature that is compatible with existing usage of container level resource/limits.
+
+
+### Non Goals
+
+
+1. This KEP focuses on the core resource types of CPU and memory. It doesn’t intend to address other resource types (e.g. GPU, storage, network bandwidth)
+at pod level in this phase. However, it could be considered in future extensions of the KEP.
+
+
+## Proposal
+
+
+### User Stories
+
+
+#### Story 1
+
+A development environment implemented as a Kubernetes Pod allows for separation of tools and a (web-based) IDE between multiple side-cars.
+The development environment defines a container with the web-server serving the IDE itself, and constrains it to use a certain amount of memory.
+Additional tools are provided in additional side-car deployments - for example an LSP( https://langserver.org/) service, a terminal, and more. 
+Using this new feature, the containers providing the terminal, the LSP services, and the set of tools being utilized can share the resource limit
+defined for the pod. Consider the following pod definition:
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  labels:
+    run: ide
+  name: myide
+spec:
+  resources:
+    limits:
+      memory: "1024Mi"
+      cpu: "4"
+  containers:
+  - image: shell
+    name: debian:buster
+  - image: tool1
+    name: first-tool:latest
+  - image: tool2
+    name: second-tool:latest
+  - name: ide
+    image: theia:latest
+    resources:
+      requests:
+        memory: 128M
+        cpu: "0.5"
+      limits:
+        memory: 256M
+        cpu: "1"
+```
+
+Using the pod-level resource definition enables the `tool1` and `tool2` containers to be constrained by the total limit for the pod. 
+
+Without this feature, the developer would need to decide a-priori how much resources to allocate to the tools - and this is not easy to do for this workload.
+
+
+#### Story 2
+
+Consider a scenario where a pod has some containers with occasional resource spikes while other containers often have idle resources. With current container 
+level resource settings, the peak usage of each container needs to be accounted for when setting limits. This leads to wasted resources when the spikes are 
+not occurring. Pod-level resource limits will allow setting a limit for the entire pod, enabling containers to share resources more efficiently without the
+need to over-provision for each container’s peak usage. 
+
+
+### Notes/Constraints/Caveats
+
+
+1. Community is trying to move cgroupv1 to maintenance mode in future Kubernetes versions (likely 1.31). Hence this feature will be supported only on cgroupv2.
+2. This feature is an **opt-in** feature, and will have no effect on existing deployments. Only deployments that explicitly require this functionality should
+turn it on by specifying the relevant resource section at pod-level in the Pod specification.
+
+
+### Risks and Mitigations
```

Comment:
another risk from the comment below is potential abuse of an issue. For example, if I have many low traffic web servers, I will likely make them best effort or bursteable. WIth Pod level requests and limits, it will be now relatively  easy to pack 20 of such containers into a single Pod and make it Guaranteed.

Yes, customer will start experience a new failure mode - noisy neighbours inside the Pod where a single container may bring the whole Pod down by exceeding limits. But the same time, there are benefits from running bigger Pods.

Exploring this abuse vector would be interesting for this KEP

Reply:
I think we can call that out in the KEP to warn users. Users should set individual container limits if they suspect that the container could potentially have high memory usage.

We can also talk about the difference in OOM behavior as well.

Reply:
OOM behavior seems like a critical aspect.  I am less worried about people over-packing a pod -- allowing users to opt-in to trouble is a core competence of Kubernetes.

And people do this already by forking too many nginx servers or whatever.  It's not really new.

Reply:
The fact that there is now a pod level limit that can be set can motivate customers to abuse it this way. As a mitigation, will this KEP be the right place to add the limit on number of containers per pod?

Reply:
> will this KEP be the right place to add the limit on number of containers per pod?

IMHO, this is not the right place to do it. As Tim mentioned, overpacking is not a new issue and instead of setting a hard limit, cluster admins can set up an admission plugin to limit this based on their own needs.

Reply:
Resolving this. Limiting the number of containers in a pod needs to be addressed in a separate KEP.

Reply:
Agree with the over-packing a pod is NOT an urgent issue to consider.